app/assets/javascripts/discourse/app/components/post/cooked-html.gjs

```diff
@@ -0,0 +1,161 @@
+import decorateMentions from "discourse/lib/post-cooked-html-decorators/mentions";
+import decorateQuoteControls from "discourse/lib/post-cooked-html-decorators/quote-controls";
+import decorateSearchHighlight from "discourse/lib/post-cooked-html-decorators/search-highlight";
+import decorateSelectionBarrier from "discourse/lib/post-cooked-html-decorators/selection-barrier";
+import { i18n } from "discourse-i18n";
+
+const detachedDocument = document.implementation.createHTMLDocument("detached");
+
+const POST_COOKED_DECORATORS = [
+  decorateSelectionBarrier,
+  decorateQuoteControls,
+  decorateLinkCounts,
+  decorateSearchHighlight,
+  decorateMentions,
+];
+
+export default class PostCookedHtml extends Component {
+  @service appEvents;
+  @service currentUser;
+
+  @tracked highlighted = false;
+  #decoratorState = new WeakMap();
+  #pendingDecoratorCleanup = [];
+
+  willDestroy() {
+    super.willDestroy(...arguments);
+    this.#cleanupDecorations();
+  }
+
+  get isStreamElement() {
+    return this.args.streamElement ?? true;
+  }
+
+  @bind
+  decorateBeforeAdopt(element, helper) {
+    this.#cleanupDecorations();
+
+    [...POST_COOKED_DECORATORS, ...this.extraDecorators].forEach(
+      (decorator) => {
+        if (!this.#decoratorState.has(decorator)) {
+          this.#decoratorState.set(decorator, {});
+        }
+
+        const decorationCleanup = decorator(element, {
+          data: {
+            post: this.post,
+            cooked: this.cooked,
+            highlightTerm: this.highlightTerm,
+            isIgnored: this.isIgnored,
+            ignoredUsers: this.ignoredUsers,
+          },
+          createDetachedElement: this.#createDetachedElement,
+          currentUser: this.currentUser,
+          helper,
+          renderNestedCookedContent: (nestedElement, cooked, extraDecorators) =>
+            helper.renderGlimmer(nestedElement, PostCookedHtml, {
+              cooked,
+              post: this.post,
+              highlightTerm: this.highlightTerm,
+              extraDecorators: [
+                ...this.extraDecorators,
+                ...makeArray(extraDecorators),
+              ],
+            }),
+          owner: getOwner(this),
+          state: this.#decoratorState.get(decorator),
+        });
+
+        if (typeof decorationCleanup === "function") {
+          this.#pendingDecoratorCleanup.push(decorationCleanup);
+        }
+      }
+    );
+
+    this.appEvents.trigger(
+      this.isStreamElement
+        ? "decorate-post-cooked-element:before-adopt"
+        : "decorate-non-stream-cooked-element",
+      element,
+      helper
+    );
+  }
+
+  @bind
+  decorateAfterAdopt(element, helper) {
+    if (this.isStreamElement) {
+      return;
+    }
+
+    this.appEvents.trigger(
+      "decorate-post-cooked-element:after-adopt",
+      element,
+      helper
+    );
+  }
+
+  get post() {
+    return this.args.data ? this.args.data.post : this.args.post;
```

Comment:
When is `this.data` ever populated? It seems like all the usages in this PR pass the model in `@post` ?

Reply:
Update the component to use `curryComponent`. 

This way, it's not necessary to use the data argument to call `renderGlimmer` to render a nested cooked component